### ISSUE TITLE
Missing XMP directory in JPG with XMP metadata

### ISSUE THREAD

**Issue by theefer** [drewnoakes/metadata-extractor#102]
It seems like XMP metadata isn't reliably extracted even when ExifTool shows XMP being present.

For instance the following image:

![image with xmp](https://raw.githubusercontent.com/drewnoakes/metadata-extractor/c7606380e0de518b2162fef2fe4130aa9911c978/Tests/Data/withXmp.jpg)

Has XMP metadata according to exiftool:

```
$ exiftool -xmp:* Tests/Data/withXmp.jpg 
XMP Toolkit                     : XMP Core 4.1.1
Date Created                    : 2015:01:22 00:00:00+00:00
...
$ exiftool -xmp:* Tests/Data/withXmp.jpg | wc -l
32
```

However doing the following returns `null`:

``` java
Metadata metadata = JpegMetadataReader.readMetadata(new File("Tests/Data/withXmp.jpg"));
Directory directory = metadata.getFirstDirectoryOfType(XmpDirectory.class);
// directory is null
```

See failing test in [my sc-missing-xmp branch](https://github.com/drewnoakes/metadata-extractor/compare/master...theefer:sc-missing-xmp), based on current master.

Is this because none of the tags present match the hardcoded tags in `XmpDirectory`? If so, is there another way to get hold of an `XmpDirectory` instance so that I can extract custom tags using `XmpDirectory.getXmpProperties()`?

**Comment by drewnoakes** [comment]
Finally had a minute to look at this.

Hex editor shows there is definitely XML data in the file. `ExtractJpegSegmentTool` shows the XMP is in the second (of two) APP1 JPEG segments. I guess the library doesn't know to look for it here. Will investigate further now and try to get a fix in tonight.

**Comment by drewnoakes** [comment]
Seems that the data in this file has a strange (invalid?) preamble:

    http://ns.adobe.com/xap/1.0/\0<?xpacket    // normal
    XMP\0://ns.adobe.com/xap/1.0/\0<?xpacket   // this file

Seems that the regular preamble has had a null-terminated `XMP` written over the top of it.

I'll patch `XmpReader` to cater for this (error?) case.

**Comment by theefer** [comment]
Great, thanks for fixing this! Would it be worth merging my PR (with tests now passing) to catch any regression of this issue?

I confess I don't know half as much as I'd need to know if this preamble is valid or not. However this issue arises with seemingly all files we get from Getty and other image agencies, so I'm erring toward this being valid somehow (not that this is always a proof of metadata quality, mind you...).

**Comment by theefer** [comment]
I noticed it wasn't actually a PR, but it is now, at https://github.com/drewnoakes/metadata-extractor/pull/104

**Comment by drewnoakes** [comment]
Interesting. If it's that common, then perhaps there's another standard I haven't come across. It works now, which is the main thing.

**Comment by theefer** [comment]
Is there any chance to get this fix published as a patch release? We have a feature that depends on being able to extract XMP metadata from such files...

On a separate note, I have been trying to add support for Getty's custom XMP namespace (which can be seen in that test file) to the library, still working on it but I'm hoping to open it as a PR at some point.

Thanks again for the fix!

**Comment by theefer** [comment]
I've further tested the latest master locally by building the JAR and loading it into our app -- all works great now!

Is there anything I could do to help get a patch release out for this?

**Comment by drewnoakes** [comment]
Good to hear. I'll get a release out in the next day or two. Hopefully
tonight.

**Comment by theefer** [comment]
Amazing, thanks a lot!

**Comment by drewnoakes** [comment]
Just released 2.8.1 to Maven Central. Should be available within a couple of hours.

**Comment by theefer** [comment]
We've released the updated library to production, works great. Thanks again!
